### PR TITLE
Add support for previous year grades

### LINKED ISSUE #99: Get grades from previous year
Similar to #73, but selecting "Previous year" in the year dropdown rather than selecting a term in the term dropdown. If this is implemented after #74, allow exporting data from the previous year.

### PR DESCRIPTION
**Pull request by psvenk**
Closes #99. Thanks to the REST API, the backend implementation here is much simpler than that in #182. While fetching grades  seems to work, I am marking this pull request as a draft because there are a few known bugs:
- [x] Assignment statistics do not work for previous-year classes
- [x] Switching between current year and previous year seems to cause some buggy behavior in the term dropdown:
image
- [x] The schedule and reports callbacks send data to `currentTableData`, which may be current or previous, meaning that the other option will not have schedule/reports data; the callbacks should be updated to send their data to both `tableData[0]` and `tableData[1]`.
- [x] Switching between years may cause the quarter displayed in the quarter/GPA dropdown to be inconsistent with the quarter whose grades are displayed. This is because, when switching years (except when switching to "previous year"), Aspine tries to look up the same quarter in the other year, which may not exist; in this case, the dropdown displays the nonexistent quarter and the table displays something else.

### CODE REVIEW

**Comment by psvenk** [comment]
> - The schedule and reports callbacks send data to `currentTableData`, which may be current or previous, meaning that the other option will not have schedule/reports data; the callbacks should be updated to send their data to both `tableData[0]` and `tableData[1]`.

Fixed in 157bc2d5f5.

**Comment by psvenk** [comment]
> - Assignment statistics do not work for previous-year classes

Fixed in cb14a441ac.

**Comment by psvenk** [comment]
> - Switching between current year and previous year seems to cause some buggy behavior in the term dropdown

Fixed in 55fa4334ca.

**Comment by psvenk** [comment]
> - Switching between years may cause the quarter displayed in the quarter/GPA dropdown to be inconsistent with the quarter whose grades are displayed. This is because, when switching years (except when switching to "previous year"), Aspine tries to look up the same quarter in the other year, which may not exist; in this case, the dropdown displays the nonexistent quarter and the table displays something else.

Fixed in 6e8037ac25.

**Comment by psvenk** [comment]
This should be ready for review. As with #220 (albeit to a lesser extent), these are fairly major changes, so it would be good to have reviews from multiple people before this is merged.

**Review by jadebuckwalter — APPROVED** [review]
Looks good to me!

**Review by MohanMaker — COMMENTED** [review]
Overall looks good to me, thanks for this great feature @psvenk! 

One bug I discovered is that you can not switch the GPA type when the previous year is selected. 

Also, the concept of cumulative GPA is a bit confusing when you are in the previous year since it still shows your cumulative GPA up to the current date. Maybe it could tell you your average GPA for that year, as well as the overall cumulative GPA?

Also, I wish it would show previous year's schedule and attendance reports, but that can come later...

**Comment by psvenk** [comment]
@jadebuckwalter @MohanMaker Thank you for your speedy reviews.

It seems that the GPA code is looking for a property in the previous year data that is not present, causing these two bugs; it should (hopefully) be a simple fix. More generally, this reminds me of @tektaxi's comment that the concept of "cumulative GPA" is more generally confusing in the previous year: if it works properly, it shows the average GPA for that year; however, the wording seems to suggest that it shows the cumulative GPA for the past two years, or for all four years, as is being tracked in #116. Maybe we can call it "yearlong GPA" or something similar?

With respect to the previous year's schedule and attendance reports, I don't think Aspen exposes this information. Maybe we should move the year dropdown so that it is clear that this only affects the "Grades" tab, but this is probably part of #187.

**Comment by psvenk** [comment]
@MohanMaker Both of the bugs that you noticed should now be fixed (the GPA type bug in 65d9a45cb6 and the cumulative GPA bug in 3a168dbbc2); could you check if this is the case on your end when you have some time?

**Review by MohanMaker — APPROVED** [review]
Just tested and everything is working correctly! I will also make a comment on the other issues you mentioned above so we can remember to discuss the dropdown and the wording of cumulative GPA.

**Review by tektaxi — APPROVED** [review]
Works well for me. I don't see any remaining bugs. We should definitely clarify the cumulative GPA thing though, right now it is still confusing.

**Pull request merged**
